File: ollama/codigo-final/fasta/llama3.1-8b/fewshot/temp_0/fasta.cpp
#include <iostream>
#include <string>
#include <vector>

// LCG constants
const int IM = 139968;
const int IA = 3877;
const int IC = 29573;
const int Seed = 42;

// Function to generate a random number using LCG
int lcg() {
    static int x = Seed;
    return (x *= (IA * x + IC)) / IM;
}

// Function to select a nucleotide based on probabilities
char selectNucleotide(const std::string& alphabet, const double* probs) {
    int r = lcg();
    for (int i = 0; i < alphabet.size(); ++i) {
        if (r < probs[i] * IM) return alphabet[i];
        r -= probs[i] * IM;
    }
    // If no nucleotide is selected, return a random one
    return 'a' + lcg() % 4;
}

// Function to generate an ALU sequence of length n*2
std::string generateALUSquence(int n) {
    std::string seq = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGA";
    for (int i = 0; i < n * 2; ++i) {
        seq += selectNucleotide("ACGT", new double[4]{0.25, 0.25, 0.25, 0.25});
    }
    return seq;
}

// Function to generate an IUB sequence of length n*3
std::string generateIUBSequence(int n) {
    std::string alphabet = "acgtBDRSVMKHN";
    double probs[] = {0.27, 0.12, 0.12, 0.27, 0.02, 0.02, 0.02, 0.02, 0.02, 0.02, 0.02, 0.02};
    std::string seq;
    for (int i = 0; i < n * 3; ++i) {
        seq += selectNucleotide(alphabet, probs);
    }
    return seq;
}

// Function to generate a Homo sapiens sequence of length n*5
std::string generateHomoSapiensSequence(int n) {
    std::string alphabet = "acgt";
    double probs[] = {0.3029549426680, 0.1979883004921, 0.1975473066391, 0.3015094502008};
    std::string seq;
    for (int i = 0; i < n * 5; ++i) {
        seq += selectNucleotide(alphabet, probs);
    }
    return seq;
}

// Function to output the sequences in the exact format
void outputSequences(int n) {
    std::cout << "ALU" << std::endl;
    for (int i = 0; i < n * 2 / 60 + 1; ++i) {
        std::string seq = generateALUSquence(n);
        for (int j = 0; j < 60 && j < seq.size(); ++j) {
            std::cout << seq[j];
        }
        if (seq.size() > 60) std::cout << std::endl;
    }

    std::cout << "Random" << std::endl;
    for (int i = 0; i < n * 3 / 60 + 1; ++i) {
        std::string seq = generateIUBSequence(n);
        for (int j = 0; j < 60 && j < seq.size(); ++j) {
            std::cout << seq[j];
        }
        if (seq.size() > 60) std::cout << std::endl;
    }

    std::cout << "Homo sapiens" << std::endl;
    for (int i = 0; i < n * 5 / 60 + 1; ++i) {
        std::string seq = generateHomoSapiensSequence(n);
        for (int j = 0; j < 60 && j < seq.size(); ++j) {
            std::cout << seq[j];
        }
        if (seq.size() > 60) std::cout << std::endl;
    }
}

int main(int argc, char* argv[]) {
    int n = atoi(argv[1]);
    outputSequences(n);
    return 0;
}

File: ollama/codigo-final/fasta/llama3.1-8b/fewshot/temp_0/fasta.py
import sys
from random import Random

# Define constants
ALU = "GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGA" \
      "TCACCTGAGGTCAGGAGTTCGAGACCAGCCTGGCCAACATGGTGAAACCCCGTCTCTACTAAAAATACAAAAATTAGCCGGGCGTGGTGGCGCGCGCCTGTAATCCCAGCTACTCGGGAGGCTGAGGCAGGAGAATCGCTTGAACCCGGGAGGCGGAGGTTGCAGTGAGCCGAGATCGCGCCACTGCACTCCAGCCTGGGCGACAGAGCGAGACTCCGTCTCAAAAA"

IUB = "acgtB*DHKMNRSVW"
HOMO_SAPIENS = "acgt"

# Define LCG parameters
IM = 139968
IA = 3877
IC = 29573
Seed = 42

def lcg(seed):
    while True:
        seed = (seed * IA + IC) % IM
        yield seed

def random(max, rng):
    return chr(rng.next() % max)

# Define IUB and Homo sapiens alphabet probabilities
IUB_PROBS = {'a': 0.27, 'c': 0.12, 'g': 0.12, 't': 0.27,
             'B': 0.02, 'D': 0.02, 'H': 0.02, 'K': 0.02,
             'M': 0.02, 'N': 0.02, 'R': 0.02, 'S': 0.02,
             'V': 0.02, 'W': 0.02}

HOMO_SAPIENS_PROBS = {'a': 0.3029549426680, 'c': 0.1979883004921,
                      'g': 0.1975473066391, 't': 0.3015094502008}

def random_fasta(alphabet, probs, rng):
    sequence = ''
    for _ in range(N * 2):
        seed = rng.next()
        char = random(ord(alphabet[seed % len(alphabet)]), rng)
        sequence += char
    return sequence

# Parse command-line argument
N = int(sys.argv[1])

# Create LCG instance
rng = lcg(Seed)

# Generate sequences
repeat_fasta = ALU * (N // 60 + 1)[:60]
random_iub_fasta = random_fasta(IUB, IUB_PROBS, rng)
random_homo_sapiens_fasta = random_fasta(HOMO_SAPIENS, HOMO_SAPIENS_PROBS, rng)

# Print output
print("Repeat Fasta")
for i in range(0, len(repeat_fasta), 60):
    print(repeat_fasta[i:i+60])

print("\nRandom IUB Fasta")
for i in range(0, len(random_iub_fasta), 60):
    print(random_iub_fasta[i:i+60])

print("\nRandom Homo sapiens Fasta")
for i in range(0, len(random_homo_sapiens_fasta), 60):
    print(random_homo_sapiens_fasta[i:i+60])
